stellargraph/layer/appnp.py

```diff
@@ -0,0 +1,244 @@
+        used by the final layer to select the predictions for those nodes in order.
+        However, the intermediate layers before the final layer order the nodes
+        in the same way as the adjacency matrix.
+
+      - The size of the final fully connected layer must be equal to the number of classes
+
+    Args:
+        layer_sizes (list of int): list of output sizes of fully connected layers in the stack
+        activations (list of str): list of activations applied to each fully connected layer's output
+        generator (FullBatchNodeGenerator): an instance of FullBatchNodeGenerator class constructed on the graph of interest
+        bias (bool): toggles an optional bias in fully connected layers
+        dropout (float): dropout rate applied to input features of each layer
+        kernel_regularizer (str): normalization applied to the kernels of fully connetcted layers
+        transport_probability: "probability" of returning to the starting node in the propogation step as desribed  in
+        the paper (alpha in the paper)
+        approx_iter: number of iterations to approximate PPNP as described in the paper (K in the paper)
+    """
+
+    def __init__(
+            self,
+            layer_sizes,
+            activations,
+            generator,
+            bias=True,
+            dropout=0.0,
+            transport_probability=0.1,
+            kernel_regularizer=None,
+            approx_iter=10
+    ):
+
+        if not isinstance(generator, FullBatchNodeGenerator):
+            raise TypeError("Generator should be a instance of FullBatchNodeGenerator")
+
+        assert len(layer_sizes) == len(activations)
+
+        self.layer_sizes = layer_sizes
+        self.transport_probability = transport_probability
+        self.activations = activations
+        self.bias = bias
+        self.dropout = dropout
+        self.kernel_regularizer = kernel_regularizer
+        self.generator = generator
+        self.support = 1
+        self.method = generator.method
+
+        # Check if the generator is producing a sparse matrix
+        self.use_sparse = generator.use_sparse
+        if self.method == "none":
+            self.graph_norm_layer = GraphPreProcessingLayer(
+                num_of_nodes=self.generator.Aadj.shape[0]
+            )
+
+        # Initialize a stack of fully connected layers
+        n_layers = len(self.layer_sizes)
+        self._layers = []
+        for ii in range(n_layers):
+            l = self.layer_sizes[ii]
+            a = self.activations[ii]
+            self._layers.append(Dropout(self.dropout))
+            self._layers.append(
+                Dense(
+                    l,
+                    activation=a,
+                    use_bias=self.bias,
+                    kernel_regularizer=self.kernel_regularizer,
+                )
+            )
+
+        feature_dim = self.layer_sizes[-1]
+
+        for ii in range(approx_iter):
+            self._layers.append(Dropout(self.dropout))
```

Comment:
Should there be a dropout between the layers of graph convolution? This is approximating the matrix inverse applied to the output activations (corresponding to the final classes we wish to classify) so I’m not sure dropout is suitable here. That being said, NN design is very empirical so we can test if there are better results with or without dropout

Reply:
Hi @adocherty, I agree with you there! The authors use dropout in between each iteration and I tried to make my implementation true to the paper. Happy to test and report back.

Reply:
Oh really? I missed that in the paper! Keep it how it's implemented then – we can test out other ways later 😄

Reply:
I did a couple of runs and the test set accuracy doesn't change so I'll remove the unnecessary dropout layers.

Reply:
I didn't see your comment above! Should I put the extra dropout layers back in?

Reply:
As you have done it now, I'd be inclined to leave them out. @PantelisElinas what do you think?

Reply:
I think we should try to be as close as possible to the paper in case a researcher uses our implementation as a baseline for comparison of their own methods against PPNP.

Reply:
I've put the extra dropout layers back in.